File: src/main/java/hm/binkley/rover/mathy/Rotation.java
package hm.binkley.rover.mathy;

import lombok.RequiredArgsConstructor;

/**
 * {@code Rotation} <b>needs documentation</b>.
 *
 * @author <a href="mailto:binkley@alumni.rice.edu">B. K. Oxley (binkley)</a>
 * @todo Needs documentation.
 */
@RequiredArgsConstructor
public enum Rotation {
    E(1, 0, 0, 1),
    N(0, -1, 1, 0),
    W(-1, 0, 0, -1),
    S(0, 1, -1, 0);

    private final int a, b, c, d;

    public Rotation left() {
        return values()[(ordinal() + 1) % 4];
    }

    public Rotation right() {
        return values()[(ordinal() + 3) % 4];
    }

    public Position prod(final Position at) {
        return Position.at(a * at.x() + b * at.y(), c * at.x() + d * at.y());
    }
}

File: src/main/kotlin/hm/binkley/rover/mathy/Rotation.kt
package hm.binkley.rover.mathy

data class Rotation(
    val a: Int,
    val b: Int,
    val c: Int,
    val d: Int,
)

val Rotation.scaling get() = a * d

operator fun Rotation.times(direction: Direction): Direction {
    return Direction(
        a * direction.x + b * direction.y,
        c * direction.x + d * direction.y,
    )
}
